bdns/dns.go

```diff
@@ -488,37 +506,51 @@ func isPrivateV6(ip net.IP) bool {
 	return false
 }
 
-func (dnsClient *impl) lookupIP(ctx context.Context, hostname string, ipType uint16) ([]dns.RR, error) {
-	resp, err := dnsClient.exchangeOne(ctx, hostname, ipType)
+func (dnsClient *impl) lookupIP(ctx context.Context, hostname string, ipType uint16) ([]dns.RR, string, error) {
+	resp, resolver, err := dnsClient.exchangeOne(ctx, hostname, ipType)
+	switch ipType {
+	case dns.TypeA:
+		if resolver != "" {
+			resolver = "A:" + resolver
+		}
+	case dns.TypeAAAA:
+		if resolver != "" {
+			resolver = "AAAA:" + resolver
+		}
+	}
 	errWrap := wrapErr(ipType, hostname, resp, err)
 	if errWrap != nil {
-		return nil, errWrap
+		return nil, resolver, errWrap
 	}
-	return resp.Answer, nil
+	return resp.Answer, resolver, nil
 }
 
 // LookupHost sends a DNS query to find all A and AAAA records associated with
 // the provided hostname. This method assumes that the external resolver will
 // chase CNAME/DNAME aliases and return relevant records. It will retry
 // requests in the case of temporary network errors. It returns an error if
 // both the A and AAAA lookups fail or are empty, but succeeds otherwise.
-func (dnsClient *impl) LookupHost(ctx context.Context, hostname string) ([]net.IP, error) {
+func (dnsClient *impl) LookupHost(ctx context.Context, hostname string) ([]net.IP, ResolverAddrs, error) {
 	var recordsA, recordsAAAA []dns.RR
 	var errA, errAAAA error
+	var resolverA, resolverAAAA string
 	var wg sync.WaitGroup
 
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		recordsA, errA = dnsClient.lookupIP(ctx, hostname, dns.TypeA)
+		recordsA, resolverA, errA = dnsClient.lookupIP(ctx, hostname, dns.TypeA)
 	}()
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		recordsAAAA, errAAAA = dnsClient.lookupIP(ctx, hostname, dns.TypeAAAA)
+		recordsAAAA, resolverAAAA, errAAAA = dnsClient.lookupIP(ctx, hostname, dns.TypeAAAA)
 	}()
 	wg.Wait()
 
+	var resolvers ResolverAddrs
+	resolvers.append(resolverA, resolverAAAA)
```

Comment:
Or better yet, since this is the only place you're actually using that `.append()` method you made:
```suggestion
	resolvers := ResolverAddrs{resolverA, resolverAAAA}
```

Reply:
I wanted to remove any empty strings from the resolvers slice. Previously I was doing [this](https://github.com/letsencrypt/boulder/pull/7271/files/758447d8a61c6c9dbd25e7f3bd3af1e69293fc31#diff-4d70665b69d43378fd22771ee9c15d3a7c5fa14dcb797aed69036ac2eb6bcb72R541-R545), but was given advice to use a custom append method. I can go back to this style, but then I'd need either filter in-line similar to the code you mentioned to remove or using my previous code with `slices.DeleteFunc`.

Reply:
FWIW I recommended that Phil just write a method since appending to this type has a requirement that the string cannot be empty. I am still partial to that approach given that now we have to either check then append twice here or loop over the resulting slice with slice.DeleteFunc when we're finished. Both of which felt worse than just writing it once as an append method that was specialized to this type.